API.md

```diff
@@ -0,0 +1,337 @@
+
+Static methods are often useful when you want to hide details on implementations.
+E.g. [Java Executors static factory method](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executors.html).
+
+You can also adopt a more flexible way to do things is to use some kind of fluent mutable builder API,
+and the invoke some kind of build method that uses all the parameters you've passed in,
+and creates some kind of immutable object.
+One of the best examples of this I've seen is [Javapoet](https://github.com/square/javapoet).
+
+I think the fluent mutable builder API is also interesting when you have to pass many parameters to construct something.
+E.g. it's a bit of pain to try to build a method invocation that takes 10 parameters.
+You have to start looking at each parameter, which type it is, the order to pass in...etc.
+With a fluent mutable builder API, you pass each parameter in a single call, in any order.
+
+The downside of fluent mutable builder API is the added complexity to support such an API.
+From a user perspective it all looks nice,
+but it's a pain to maintain it.
+
+My feeling is that LDK fits somewhere between static factory methods and fluent builder API.
+Some of its static factory methods take a lot of parameters.
+If that can be reduced somehow then I'd favour static factory methods.
+As an example: `ChainMonitor` static factory takes both `tx_broadcaster` and `fee_estimator`.
+Then, when constructing `ChannelManager` you pass in those two parameters,
+as well as as a `ChainMonitor` as a watch.
+Maybe things can be tidied up to have more manageable static factory methods.
+
+Also, I would often use words like `of` or `from` rather than `constructor_new`.
+E.g. `PeerManager.of()`.
+
+
+## KeysManager constructor parameters
+
+Aside from where the time comes from,
+I noticed that KeysManager constructor takes the current time in both secs and nanoseconds.
+That sounds odd.
+Why do you need both parameters?
+Can't you derive the seconds from the nanoseconds
+(e.g. using `TimeUnit.NANOSECONDS.toSeconds(nanos)`) ? 
+
+
+## Key seed management
+
+In initializing `KeysManager` [section](https://lightningdevkit.org/docs/build_node#initialize-the-keysmanager),
+it says:
+
+> Note that you must write the key_seed you give to the KeysManager on startup to disk,
+> and keep using it to initialize the KeysManager every time you restart.
+> This key_seed is used to derive your node's secret key (which corresponds to its node pubkey) and all other secret key material.
+
+I was wondering, can't LDK take of this?
+It can be hooked with a persister, 
+which could potentially persist it and restore it on restart.
+The comment above seems to indicate that it's the user's responsibility to write it to disk and restore it.
+
+
+## Fresh Start vs Restart
+
+While reading [the guide](https://lightningdevkit.org/docs/build_node/),
```

Comment:
Hmmm, I mean I suppose `ChannelManagerConstructor` could take nulls for the serialized stuff that was loaded from disk.